clash-prelude/src/Clash/Class/Exp.hs

```diff
@@ -0,0 +1,93 @@
+{-|
+Copyright  :  (C) 2019, Myrtle Software Ltd
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>
+-}
+
+{-# LANGUAGE CPP           #-}
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE MagicHash     #-}
+{-# LANGUAGE TypeFamilies  #-}
+{-# LANGUAGE TypeOperators #-}
+
+#if __GLASGOW_HASKELL__ >= 806
+{-# LANGUAGE NoStarIsType  #-}
+#endif
+
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
+
+module Clash.Class.Exp (Exp, ExpResult, (^)) where
+
+import qualified Prelude                       as P
+import           Prelude                       hiding ((^))
+
+import           Clash.Annotations.Primitive   (hasBlackBox)
+import           Clash.Promoted.Nat            (SNat(..), snatToInteger)
+import           Clash.Sized.Internal.Index    (Index)
+import           Clash.Sized.Internal.Signed   (Signed)
+import           Clash.Sized.Internal.Unsigned (Unsigned)
+
+import           GHC.TypeLits
+  (KnownNat, Nat, type (^), type (*))
+
+-- | Type class implementing exponentiation with explicitly resizing results.
+class Exp a where
+  type ExpResult a (n :: Nat)
+
+  -- | Exponentiation with known exponent.
+  (^)
+    :: KnownNat n
```

Comment:
Additional `KnownNat n` should not be needed because you already have an `SNat n`. So if you have a function:

```
f :: SNat -> a -> r
f sn x = ...
```

and GHC complains about not having an `KnownNat n` available somewhere in the context of `f`, then simply change the definition of f to:

```
f :: SNat -> a -> r
f sn@SNat x = ...
```

i.e. pattern-match on the `SNat` constructor to bring the `KnownNat` constraint into scope.

Reply:
Something you already seem to be doing in all of your instances... so I'm not sure why the class method has this `KnownNat n` constraint?

Reply:
Ah darn, I guess that's a left-over from when I was using `Unsigned n`! Thanks!